subsys/net/lib/ocpp/core.c

```diff
@@ -0,0 +1,504 @@
+	rmsg.msg_len = strlen(buf) + 1;
+	rmsg.sndlock = &ui->ws_sndlock;
+	rmsg.rspsig = &ui->ws_rspsig;
+	ret = ocpp_send_to_server(&rmsg, K_MSEC(timeout_ms));
+	if (ret < 0) {
+		return ret;
+	}
+
+	*status = sh->resp_status;
+	if (sh->resp_status == OCPP_AUTH_ACCEPTED) {
+		sh->is_active = true;
+	}
+
+	return 0;
+}
+
+int ocpp_heartbeat(ocpp_session_handle_t hndl)
+{
+	int ret;
+	struct ocpp_session *sh = (struct ocpp_session *)hndl;
+	struct ocpp_info *ctx = sh->ctx;
+	struct ocpp_upstream_info *ui = ctx->ui;
+	ocpp_msg_fp_t fn;
+	struct ocpp_wamp_rpc_msg rmsg = {0};
+	char *buf = ctx->pdu_buf;
+
+	fn = ctx->cfn[PDU_HEARTBEAT];
+	sh->uid = fn(buf, sizeof(ctx->pdu_buf), sh);
+
+	rmsg.ctx = ctx;
+	rmsg.msg = buf;
+	rmsg.msg_len = strlen(buf) + 1;
+	rmsg.sndlock = &ui->ws_sndlock;
+	rmsg.rspsig = &ui->ws_rspsig;
+
+	ret = ocpp_send_to_server(&rmsg, K_SECONDS(OCPP_PDU_TIMEOUT));
+	if (!ret && sh->resp_status != BOOT_ACCEPTED) {
+		ret = -EAGAIN;
+	}
+
+	return ret;
+}
+
+int ocpp_start_transaction(ocpp_session_handle_t hndl,
+			   int Wh,
+			   uint8_t conn_id,
+			   uint32_t timeout_ms)
+{
+	struct ocpp_session *sh = (struct ocpp_session *)hndl;
+	struct ocpp_info *ctx;
+	struct ocpp_upstream_info *ui;
+	char buf[350];
+	struct ocpp_wamp_rpc_msg rmsg = {0};
+	ocpp_msg_fp_t fn;
+	char utc[CISTR25] = {0};
+	int ret;
+
+	if (!ocpp_session_is_valid(sh) ||
+	    conn_id <= 0) {
+		return -EINVAL;
+	}
+	ctx = sh->ctx;
+	sh->idcon = conn_id;
+	if (ctx->state < CP_STATE_READY) {
+		return -EAGAIN;
+	}
+
+	if (ctx->is_cs_offline) {
+		/* todo: fill meter reading
+		 * server offline, accept start txn and save it in Q.
+		 */
+		return 0;
+	}
+
+	fn = ctx->cfn[PDU_START_TRANSACTION];
+	ocpp_get_utc_now(utc);
+	sh->uid = fn(buf, sizeof(buf), sh, Wh, -1, utc);
+
+	ui = ctx->ui;
+	rmsg.ctx = ctx;
+	rmsg.msg = buf;
+	rmsg.msg_len = strlen(buf) + 1;
+	rmsg.sndlock = &ui->ws_sndlock;
+	rmsg.rspsig = &ui->ws_rspsig;
+	ret = ocpp_send_to_server(&rmsg, K_MSEC(timeout_ms));
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (sh->resp_status != OCPP_AUTH_ACCEPTED) {
+		sh->is_active = false;
+		ret = -EACCES;
+	} else {
+		union ocpp_keyval *keyval;
+
+		keyval = ocpp_get_key_val(CFG_MTR_VAL_SAMPLE_INTERVAL);
+		if (!atomic_inc(&ctx->mtr_timer_ref_cnt)) {
+			k_timer_start(&ctx->mtr_timer,
+				      K_SECONDS(keyval->ival),
+				      K_SECONDS(keyval->ival));
+		}
+	}
+
+	return ret;
+}
+
+int ocpp_stop_transaction(ocpp_session_handle_t hndl,
+			  int Wh,
+			  uint32_t timeout_ms)
+{
+	struct ocpp_session *sh = (struct ocpp_session *)hndl;
+	struct ocpp_info *ctx;
+	struct ocpp_upstream_info *ui;
+	char buf[350];
+	char utc[CISTR25] = {0};
+	struct ocpp_wamp_rpc_msg rmsg = {0};
+	ocpp_msg_fp_t fn;
+	int ret;
+
+	if (!ocpp_session_is_valid(sh)) {
+		return -EINVAL;
+	}
+
+	ctx = sh->ctx;
+	if (ctx->state < CP_STATE_READY) {
+		return -EAGAIN;
+	}
+
+	sh->is_active = false;
+	if (atomic_dec(&ctx->mtr_timer_ref_cnt) <= 1) {
+		k_timer_stop(&ctx->mtr_timer);
+	}
```

Comment:
For reference counting you should probably use `atomic_cas()` function instead.  Even if `atomic_dec` is atomic, the check after it is not. It would also make sense to create a separate ref/unref functions that would deal all the nasty details of reference counting.
See for example `subsys/net/ip/net_pkt.c` and `net_pkt_unref()` and `net_pkt_ref()` how it is done there.

Reply:
the logical comparison of ref count is done after atomic_cas in subsys/net/ip/net_pkt.c, the same is used here atomic_dec. I don't find the if (ref <= 1) can be achieved using atomic_cas

 544         } while (!atomic_cas(&pkt->atomic_ref, ref, ref - 1));
 545 

 552         if (ref > 1) {
 553                 goto done;
 554         }